### PR TITLE
Add ban count option to round editor

### PR DESCRIPTION
**Pull request by rodrigopina360**
# osu.Game.Tournament.Models
+ Add: New property BanCount in TournamentRound to save the number of bans

# osu.Game.Tournament/Screens
+ Add: New slider setting in RoundEditorScreen to select the number of bans per round
* Change: Modified setNextMode behavior to get the round ban count, and select bans accordingly

~ This idea came from this [comment](https://github.com/ppy/osu/issues/24857#issuecomment-1725083733), which remembered me that indeed it's always assumed it's a single ban, even tho (even in OWC) having 2 bans per team further in the bracket is the norm. This development does just that, allowing for the number of bans to be specified per round. I decided in a 0-5 ban count to be on the safe side.

### CODE REVIEW

**Comment by peppy** [comment]
I've tidied this up a bit. Waiting on a response from OWC stream management to see how they are currently working around this before merging.

**Comment by peppy** [comment]
@rodrigopina360 for official world cups, it would seem the ban order is always `ABBA` – does that also hold for your use cases? And would this by extension mean `ABBAABBA` for 4 bans, for example?

**Comment by rodrigopina360** [comment]
> @rodrigopina360 for official world cups, it would seem the ban order is always `ABBA` – does that also hold for your use cases? And would this by extension mean `ABBAABBA` for 4 bans, for example?

Yeah, that was a bit of an oversight from my end, right now it just goes ABAB, will handle that later in the day.

**Comment by rodrigopina360** [comment]
Issue fixed, the ban order logic now takes in consideration the # of total bans, and gives 2 consecutive team bans after the first ban until all bans have been performed, or in short, using the ABBA (ABBAAB or ABBAABBA) ban order.

File: osu.Game.Tournament/Screens/MapPool/MapPoolScreen.cs
Comment (bdach) [inline osu.Game.Tournament/Screens/MapPool/MapPoolScreen.cs:172]: I feel like this is written in a very roundabout and difficult-to-follow manner... I would much prefer something like the following:

```csharp
            var previousColour = CurrentMatch.Value.PicksBans.LastOrDefault()?.Team ?? roll_winner;
            TeamColour nextColour;

            bool hasAllBans = CurrentMatch.Value.PicksBans.Count(p => p.Type == ChoiceType.Ban) >= totalBansRequired;

            if (!hasAllBans)
            {
                // Ban phase.
                // Switch teams every second ban.
                nextColour = CurrentMatch.Value.PicksBans.Count % 2 == 1
                    ? getOppositeTeamColour(previousColour)
                    : previousColour;
            }
            else if (pickType == ChoiceType.Ban)
            {
                // Switching from bans to picks - stay with the last team that was banning.
                nextColour = pickColour;
            }
            else
            {
                // Pick phase.
                // Switch teams every pick.
                nextColour = getOppositeTeamColour(previousColour);
            }

            setMode(nextColour, hasAllBans ? ChoiceType.Pick : ChoiceType.Ban);

            TeamColour getOppositeTeamColour(TeamColour colour) => colour == TeamColour.Red ? TeamColour.Blue : TeamColour.Red;
```

Note that the above is untested. (This PR really should come with tests at this point too.)

**Comment by eclecticfluff** [comment]
It might be too much of a hassle to implement, especially this close to being merged, but the [Thue–Morse sequence](https://en.m.wikipedia.org/wiki/Thue–Morse_sequence#Equitable_sequencing) would be better than both ABABAB and ABBAAB ordering.

**Comment by peppy** [comment]
> It might be too much of a hassle to implement, especially this close to being merged, but the [Thue–Morse sequence](https://en.m.wikipedia.org/wiki/Thue%E2%80%93Morse_sequence#Equitable_sequencing) would be better than both ABABAB and ABBAAB ordering.

please take that to the tournament organisers, not here.

**Comment by rodrigopina360** [comment]
Okay so, to make sure I got it, I should simplify the code to be more readable, like @bdach suggested, and add in a few tests to make sure everything is running smoothly right? I've been dabling with tests before doing the PR, but I failed to find a way to programatically click the mappool slot to make sure the correct action was performed. I'll investigate it further.

**Review by bdach — DISMISSED** [review]

**Review by peppy — APPROVED** [review]

**Pull request merged**
